python/lsst/ip/isr/isrMock.py

```diff
@@ -0,0 +1,915 @@
+        ----------
+        ampData : `lsst.afw.image.ImageF`
+            Amplifier image to operate on.
+        start : float
+            Start value of the gradient (at y=0).
+        end : float
+            End value of the gradient (at y=ymax).
+        """
+        nPixY = ampData.getDimensions().getY()
+        ampArr = ampData.getArray()
+        ampArr[:] = ampArr[:] + (np.interp(range(nPixY), (0, nPixY - 1), (start, end)).reshape(nPixY, 1) +
+                                 np.zeros(ampData.getDimensions()).transpose())
+
+    def amplifierAddSource(self, ampData, scale):
+        r"""Add a single Gaussian source to an amplifier.
+
+         This method operates in the amplifier coordinate frame.
+
+        Parameters
+        ----------
+        ampData : `lsst.afw.image.ImageF`
+            Amplifier image to operate on.
+        scale : float
+            Peak flux of the source to add.
+        """
+        for x in range(0, ampData.getDimensions().getX()):
+            for y in range(0, ampData.getDimensions().getY()):
+                ampData.getArray()[y][x] = (ampData.getArray()[y][x] +
+                                            scale * np.exp(-0.5 * ((x - 50)**2 + (y - 25)**2)**2 / 3.0**2))
+
+    def amplifierAddCT(self, ampDataSource, ampDataTarget, scale):
+        r"""Add a scaled copy of an amplifier to another, simulating crosstalk.
+
+         This method operates in the amplifier coordinate frame.
+
+        Parameters
+        ----------
+        ampData : `lsst.afw.image.ImageF`
+            Amplifier image to operate on.
+        scale : float
+            Peak flux of the source to add.
+
+        Notes
+        -----
+        This simulates simple crosstalk between amplifiers.
+        """
+        for x in range(0, ampDataSource.getDimensions().getX()):
+            for y in range(0, ampDataSource.getDimensions().getY()):
+                ampDataTarget.getArray()[y][x] = (ampDataTarget.getArray()[y][x] +
+                                                  scale * ampDataSource.getArray()[y][x])
+
+    # Functional form data values
+    def amplifierAddFringe(self, amp, ampData, scale):
+        r"""Add a fringe-like ripple pattern to an amplifier's image data.
+
+        Parameters
+        ----------
+        amp : `lsst.afw.ampInfo.AmpInfoRecord`
+            Amplifier to operate on.  Needed for amp<->exp coordinate transforms.
+        ampData : `lsst.afw.image.ImageF`
+            Amplifier image to operate on.
+        scale : float
+            Peak intensity scaling for the ripple.
+
+        Notes
+        -----
+        This uses an offset sinc function to generate a ripple
+        pattern.  True fringes have much finer structure, but this
+        pattern should be visually identifiable.  The (x, y)
+        coordinates are in the frame of the amplifier, and (u, v) in
+        the frame of the full trimmed image.
+        """
+        for x in range(0, ampData.getDimensions().getX()):
```

Comment:
I guess I'll let you off the pixel loop here, because I can't be bothered to work out the alternative because it's gone 1am here. However, there are some other problems. 

The sinc function rings down very fast, and the `/50` isn't really enough to keep it alive over much of the chip at all, given it goes 4k in both directions for LSST. Perhaps something a lot larger would be better? But I do realise that will look very silly near the center of the sinc though...

Also though, while I do see the similarity between `sinc` and fringes, perhaps there are better options for this. Might a simple 2d sin wave look better?

Reply:
Also, again, center of the fringe is very close to chip edge. Moving it in might help things a bit too.

Reply:
Oh, just noticed - it's moved in on one axis, but is literally on the edge on the other? Shouldn't something be being subtracted from `v` as well as `u`? Or did you really want the middle of the fringe on the literal edge?

Reply:
This functional form is also arbitrary, and although many other forms would work, this met all of my initial criteria:

- Easily recognizable by eye (the arcs are obviously artificial, and are also not straight lines).
- Small number of free parameters (although DM-4441 breaks this to allow multiple fringe patterns; that will allow the center and flux scale to be arrays of floats).
- Varying intensity of signal (I don't think this is exercised the way I wanted, but I wanted the signal to be large in some amplifiers and small in others to check that the correction deals well with that).

Reply:
That all sounds very reasonable. I still think that it decays too quickly to meet your last spec though, and the signal will be undetectable in most amps without putting an absurdly large signal down at the centre. Maybe just tweak the 50 pixel length scale in this case?

Reply:
I've increased this, and will think about making it a parameter in the multi-fringe testing ticket (DM-4441).